consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/ValidatorContractTest.java

```diff
@@ -121,6 +126,122 @@ public void migratesFromBlockToValidatorContract() {
     assertThat(extraDataCodec.decode(block1).getVote()).isEmpty();
   }
 
+  @Test
+  public void transitionsFromValidatorContractModeToBlockHeaderMode() {
+    final List<QbftFork> qbftForks = List.of(createBlockHeaderFork(1));
+    final TestContext context =
+        new TestContextBuilder()
+            .indexOfFirstLocallyProposedBlock(0)
+            .nodeParams(
+                List.of(new NodeParams(NODE_ADDRESS, NodeKeyUtils.createFrom(NODE_PRIVATE_KEY))))
+            .clock(fixedClock)
+            .genesisFile(
+                Resources.getResource("genesis_migrating_validator_blockheader.json").getFile())
+            .useValidatorContract(true)
+            .qbftForks(qbftForks)
+            .buildAndStart();
+
+    // block 0 uses validator contract with 1 validator
+    // block 1 onwards uses block header voting (which reuses previous block's validators upon
+    // switching)
+    final List<Address> block0Addresses = Stream.of(NODE_ADDRESS).collect(Collectors.toList());
+
+    createNewBlockAsProposer(context, 1);
+
+    final ValidatorProvider validatorProvider = context.getValidatorProvider();
+    final BlockHeader genesisBlock = context.getBlockchain().getBlockHeader(0).get();
+    final BlockHeader block1 = context.getBlockchain().getBlockHeader(1).get();
+
+    // contract block extra data cannot contain validators or vote
+    assertThat(validatorProvider.getValidatorsForBlock(genesisBlock)).isEqualTo(block0Addresses);
+    assertThat(extraDataCodec.decode(genesisBlock).getValidators()).isEmpty();
+    assertThat(extraDataCodec.decode(genesisBlock).getVote()).isEmpty();
+
+    // uses previous block's validators
+    assertThat(validatorProvider.getValidatorsForBlock(block1)).isEqualTo(block0Addresses);
+    assertThat(extraDataCodec.decode(block1).getValidators()).containsExactly(NODE_ADDRESS);
```

Comment:
Wonder if there is anything we can do to prove that the contract wasn't called to obtain the validators